btree/btree.c

```diff
@@ -0,0 +1,13031 @@
+	COPY_VALUE(&node_slot.s_rec, &bop->bo_rec);
+	/* if we need to update vec_count for root slot, update at this place */
+
+	bnode_done(&node_slot, true);
+	bnode_seq_cnt_update(lev->l_node);
+	bnode_fix(lev->l_node);
+	/**
+	 * Note : not capturing l_node as it must have already been captured in
+	 * btree_put_makespace_phase().
+	 */
+	btree_node_capture_enlist(oi, oi->i_extra_node, 0);
+
+	/* Increase height by one */
+	tree->t_height++;
+	bop->bo_arbor->t_height = tree->t_height;
+	if (tree->t_height > MAX_TREE_HEIGHT) {
+		M0_LOG(M0_ERROR,
+		       "Tree height increased beyond MAX_TREE_HEIGHT");
+		M0_ASSERT(0);
+	}
+	/* Capture this change in transaction */
+
+	/* TBD : This check needs to be removed when debugging is done. */
+	M0_ASSERT_EX(bnode_expensive_invariant(lev->l_node));
+	M0_ASSERT_EX(bnode_expensive_invariant(oi->i_extra_node));
+	bnode_unlock(lev->l_node);
+	bnode_unlock(oi->i_extra_node);
+
+	return P_CAPTURE;
+}
+
+/**
+ * This function is called when there is overflow and splitting needs to be
+ * done. It will move some records from right node(l_node) to left node(l_alloc)
+ * and find the appropriate slot for given record. It will store the node and
+ * index (where we need to insert given record) in tgt slot as a result.
+ *
+ * @param l_alloc It is the newly allocated node, where we want to move record.
+ * @param l_node It is the current node, from where we want to move record.
+ * @param rec It is the given record for which we want to find slot
+ * @param tgt result of record find will get stored in tgt slot
+ */
+static void btree_put_split_and_find(struct nd *allocated_node,
+				     struct nd *current_node,
+				     struct m0_btree_rec *rec, struct slot *tgt)
+{
+	struct slot              right_slot;
+	struct slot              left_slot;
+	struct m0_bufvec_cursor  cur_1;
+	struct m0_bufvec_cursor  cur_2;
+	int                      diff;
+	m0_bcount_t              ksize;
+	void                    *p_key;
+	m0_bcount_t              vsize;
+	void                    *p_val;
+	int                      min_rec_count;
+
+	/* intialised slot for left and right node*/
+	left_slot.s_node  = allocated_node;
+	right_slot.s_node = current_node;
+
+	/* 1)Move some records from current node to new node */
+
+	bnode_set_level(allocated_node, bnode_level(current_node));
+
+	bnode_move(current_node, allocated_node, D_LEFT, NR_EVEN);
+
+	/**
+	 * Assert that nodes still contain minimum number of records in the node
+	 * required by btree. If Assert fails, increase the node size or
+	 * decrease the object size.
+	 */
+	min_rec_count = bnode_level(current_node) ? 2 : 1;
+	M0_ASSERT(bnode_count_rec(current_node) >= min_rec_count);
+	M0_ASSERT(bnode_count_rec(allocated_node) >= min_rec_count);
+	/*2) Find appropriate slot for given record */
+
+	right_slot.s_idx = 0;
+	REC_INIT(&right_slot.s_rec, &p_key, &ksize, &p_val, &vsize);
+	bnode_key(&right_slot);
+
+	if (current_node->n_tree->t_keycmp.rko_keycmp != NULL) {
```

Comment:
[Suggestion] Add a helper function function for comparing the key data. 

diff = bnode_key_cmp(data1, data2)

bnode_key_cmp()
{

	if (current_node->n_tree->t_keycmp.rko_keycmp != NULL) {
		diff = current_node->n_tree->t_keycmp.rko_keycmp(
				M0_BUFVEC_DATA(&rec->r_key.k_data),
				M0_BUFVEC_DATA(&right_slot.s_rec.r_key.k_data));
	} else {
		m0_bufvec_cursor_init(&cur_1, &rec->r_key.k_data);
		m0_bufvec_cursor_init(&cur_2, &right_slot.s_rec.r_key.k_data);

		diff = m0_bufvec_cursor_cmp(&cur_1, &cur_2);
	}
	
	return diff;
}

Reply:
The activity of removal of all duplicated key compare code from btree will be taken separately later, please consider and approve this for now.